### PR TITLE
M7: Skill Builder

### LINKED ISSUE #21: [FEATURE] Design Builder data model
Define builder data types in `src/builder.rs`:

```rust
/// User input for skill generation
#[derive(Debug, Clone)]
pub struct SkillSpec {
    pub purpose: String,
    pub name: Option<String>,
    pub tools: Option<String>,
    pub compatibility: Option<String>,
    pub license: Option<String>,
    pub extra_files: Option<HashMap<String, String>>,
}

/// Result of skill generation
#[derive(Debug)]
pub struct BuildResult {
    pub properties: SkillProperties,
    pub files: HashMap<String, String>,  // relative-path → content
    pub output_dir: PathBuf,
}

/// Clarity assessment result
#[derive(Debug)]
pub struct ClarityAssessment {
    pub clear: bool,
    pub questions: Vec<String>,
}
```

### LINKED ISSUE #22: [FEATURE] Implement Builder core
Implement builder core logic in `src/builder.rs`:

- `build_skill(spec: &SkillSpec, output_dir: &Path) -> Result<BuildResult>` — generates SKILL.md with valid frontmatter + markdown body, creates directory structure, writes files, validates result
- `derive_name(purpose: &str) -> String` — derives kebab-case skill name from natural language description (prefer gerund form: "Process PDFs" → "processing-pdfs")
- `generate_frontmatter(spec: &SkillSpec) -> HashMap<String, serde_yaml_ng::Value>` — builds YAML frontmatter
- `generate_body(spec: &SkillSpec) -> String` — generates markdown body (prompt content)
- `assess_clarity(purpose: &str) -> ClarityAssessment` — evaluates if a one-liner is clear enough to generate a skill autonomously, or returns clarifying questions

**Spec compliance (from Anthropic best practices):**
- Post-creation validation: run `validate()` on generated output

### LINKED ISSUE #23: [FEATURE] Builder: LLM-supported skill generation
The Builder should support two generation modes:

### Mode 1: Deterministic (default, no configuration needed)
- Uses heuristics, templates, and string manipulation
- Produces valid but formulaic output
- Always available — zero external dependencies

### Mode 2: LLM-enhanced (when API key is configured)
- Uses an LLM (e.g., Claude) to generate richer descriptions, bodies, and names
- Produces higher-quality, context-aware output
- Requires `ANTHROPIC_API_KEY` environment variable

### Implementation
- Use `reqwest` for HTTP calls to the Anthropic API
- Mode selection: automatic based on `ANTHROPIC_API_KEY`, or explicit via `--no-llm`
- Fallback: if LLM call fails, fall back to deterministic mode with a warning to stderr

### Affected functions
- `derive_name`: LLM produces more natural gerund names
- `generate_description`: LLM produces richer third-person what+when descriptions
- `generate_body`: LLM generates detailed, context-specific instructions
- `assess_clarity`: LLM provides better ambiguity detection

### Acceptance criteria
- Builder works with no configuration (deterministic mode)
- Setting `ANTHROPIC_API_KEY` enables LLM-enhanced generation
- `--no-llm` flag forces deterministic mode
- LLM failures gracefully fall back to deterministic mode

### LINKED ISSUE #24: [FEATURE] Builder: support multiple LLM providers and local models
Expand LLM support beyond Anthropic to all major providers.

### Providers to support

| Provider | API endpoint | Env var | Notes |
|----------|-------------|---------|-------|
| Anthropic (Claude) | `api.anthropic.com/v1/messages` | `ANTHROPIC_API_KEY` | Current default |
| OpenAI | `api.openai.com/v1/chat/completions` | `OPENAI_API_KEY` | GPT-4, etc. |
| Google (Gemini) | `generativelanguage.googleapis.com` | `GOOGLE_API_KEY` | Gemini Pro, etc. |
| Local (Ollama) | `localhost:11434/api/generate` | (none needed) | No API key required |
| OpenAI-compatible | Configurable base URL | Configurable | Covers vLLM, LM Studio, etc. |

### Design: trait-based abstraction

```rust
#[async_trait]
pub trait LlmProvider: Send + Sync {
    async fn generate(&self, system: &str, user: &str) -> Result<String>;
}
```

Each provider implements this trait. Provider selection via:
- `--provider` CLI flag or `AIGENT_LLM_PROVIDER` env var
- Auto-detection: check env vars in priority order
- `--no-llm` forces deterministic mode

### Provider detection order
1. `ANTHROPIC_API_KEY` → Anthropic
2. `OPENAI_API_KEY` → OpenAI
3. `GOOGLE_API_KEY` → Google
4. Check `localhost:11434` → Ollama
5. None → deterministic mode

Depends on LLM support issue.

### LINKED ISSUE #25: [TEST] Write Builder tests
Write builder tests:

- `derive_name` from natural language descriptions (gerund form)
- `generate_frontmatter` from SkillSpec
- Full `build_skill` pipeline (creates valid directory with SKILL.md)
- Validation of generated output passes (`validate()` returns empty errors)
- `assess_clarity` — clear vs ambiguous purposes
- Edge cases: long descriptions, special characters in purpose
- `init` creates template SKILL.md
- Rejects building into existing directory
- LLM mode: mock/stub strategy for deterministic tests (use `mockall` or similar)

### PR DESCRIPTION
**Pull request by wkusnierczyk**
## Summary

- Refactor `src/builder.rs` into `src/builder/` module directory with `deterministic.rs`, `llm.rs`, `template.rs`, and `providers/` submodule
  Closes #21
- Implement deterministic skill builder: name derivation (gerund-form, kebab-case), description and body generation (template-based), clarity assessment (heuristic)
  Closes #22
- Add `LlmProvider` trait with env-var-based `detect_provider()` and per-function LLM-with-fallback pattern in `build_skill`
  Closes #23
- Implement four LLM backends (Anthropic, OpenAI, Google, Ollama) using `ureq` sync HTTP
  Closes #24
- Add 51 new tests (23 deterministic, 6 LLM mock, 15 builder unit, 7 CLI integration) — 169 total
  Closes #25

## Test plan

- [x] `cargo fmt --check` — clean
- [x] `cargo clippy -- -D warnings` — clean
- [x] `cargo test` — 169 tests pass (146 unit + 23 integration)
- [x] `cargo build --release` — success
- [x] Smoke test: `aigent build "Process PDF files" --no-llm` creates valid skill
- [x] Smoke test: `aigent init <dir>` creates template SKILL.md
- [x] Round-trip: built skill passes `aigent validate`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/builder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    }
+
+    // 12. Validate output.
+    let messages = validate(&output_dir);
+    let errors: Vec<&str> = messages
+        .iter()
+        .filter(|m| !m.starts_with("warning: "))
+        .map(|m| m.as_str())
+        .collect();
+    if !errors.is_empty() {
```
[the remark this answers is not in the document]
Comment (wkusnierczyk) [reply src/builder/mod.rs]: Fixed in f20f382. `build_skill` now cleans up the written SKILL.md and any extra files before returning the validation error, so no orphaned artifacts remain on failure.

File: src/builder/deterministic.rs [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+        format!("{capitalized}.")
+    };
+
+    // Derive trigger context from purpose.
+    let trigger = derive_trigger(purpose);
+    let description = format!("{sentence} Use when {trigger}.");
+
+    // Truncate to 1024 characters if needed.
+    if description.len() > 1024 {
+        description[..1024].to_string()
```
[the remark this answers is not in the document]
Comment (wkusnierczyk) [reply src/builder/deterministic.rs]: Fixed in f20f382. Now uses `description.chars().take(1024).collect()` for safe UTF-8 truncation, matching the same fix applied in `llm.rs`.

File: src/builder/mod.rs [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+        }
+    } else {
+        deterministic::derive_name(&spec.purpose)
+    };
+
+    // 2. Determine output directory.
+    let output_dir = spec
+        .output_dir
+        .clone()
+        .unwrap_or_else(|| PathBuf::from(&name));
```
[the remark this answers is not in the document]
Comment (wkusnierczyk) [reply src/builder/mod.rs]: This is by design. When `spec.name` is not provided, the name is derived from `purpose` (via LLM or deterministic heuristics) to give a meaningful default. The output directory and skill name serve different roles — the directory is a filesystem concern, while the name is a semantic identifier in the frontmatter. The validator independently validates the final name against spec rules (kebab-case, length, allowed characters), so invalid derived names are caught at the validation step. All tests use matching name/dir pairs because they test the generation logic, not the mismatch case.

File: src/builder/mod.rs [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    files.insert("SKILL.md".to_string(), content);
+
+    if let Some(ref extra) = spec.extra_files {
+        for (rel_path, file_content) in extra {
+            let full_path = output_dir.join(rel_path);
+            if let Some(parent) = full_path.parent() {
+                std::fs::create_dir_all(parent)?;
+            }
+            std::fs::write(&full_path, file_content)?;
+            files.insert(rel_path.clone(), file_content.clone());
```
[the remark this answers is not in the document]
Comment (wkusnierczyk) [reply src/builder/mod.rs]: Fixed in f20f382. Added path traversal validation that rejects absolute paths and any paths containing `".."` components before joining onto `output_dir`. Returns an `AigentError::Build` with a descriptive message.
